File: modules/core/src/main/kotlin/org/apache/utlx/core/udm/UDMLanguageSerializer.kt
package org.apache.utlx.core.udm

import java.time.format.DateTimeFormatter

/**
 * Serializes UDM structures to UDM Language format (.udm files)
 *
 * UDM Language is a meta-format that preserves complete UDM model state including:
 * - Type information (Scalar, Array, Object, DateTime, etc.)
 * - Metadata (source info, line numbers, validation state)
 * - Attributes (XML attributes, hints)
 * - Element names (XML context)
 *
 * This is different from standard YAML/JSON serialization which loses UDM metadata.
 *
 * Example:
 * ```kotlin
 * val udm = UDM.Object(...)
 * val udmLang = UDMLanguageSerializer().serialize(udm)
 * File("output.udm").writeText(udmLang)
 * ```
 */
class UDMLanguageSerializer(
    private val prettyPrint: Boolean = true,
    private val indentSize: Int = 2
) {

    /**
     * Serialize UDM to UDM Language format
     */
    fun serialize(udm: UDM, sourceInfo: Map<String, String> = emptyMap()): String {
        val sb = StringBuilder()

        // Header
        sb.append("@udm-version: 1.0\n")
        if (sourceInfo.isNotEmpty()) {
            sourceInfo["source"]?.let { sb.append("@source: \"$it\"\n") }
            sourceInfo["parsed-at"]?.let { sb.append("@parsed-at: \"$it\"\n") }
        }
        sb.append("\n")

        // Body
        serializeValue(udm, sb, 0)

        return sb.toString()
    }

    private fun serializeValue(udm: UDM, sb: StringBuilder, depth: Int) {
        when (udm) {
            is UDM.Scalar -> serializeScalar(udm, sb)
            is UDM.Array -> serializeArray(udm, sb, depth)
            is UDM.Object -> serializeObject(udm, sb, depth)
            is UDM.DateTime -> sb.append("@DateTime(\"${udm.instant}\")")
            is UDM.Date -> sb.append("@Date(\"${udm.date}\")")
            is UDM.LocalDateTime -> sb.append("@LocalDateTime(\"${udm.dateTime.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)}\")")
            is UDM.Time -> sb.append("@Time(\"${udm.time.format(DateTimeFormatter.ISO_LOCAL_TIME)}\")")
            is UDM.Binary -> serializeBinary(udm, sb)
            is UDM.Lambda -> sb.append("@Lambda()")
        }
    }

    private fun serializeScalar(scalar: UDM.Scalar, sb: StringBuilder) {
        when (val value = scalar.value) {
            null -> sb.append("null")
            is String -> sb.append("\"${escapeString(value)}\"")
            is Boolean -> sb.append(value.toString())
            is Number -> sb.append(value.toString())
            else -> {
                // Explicit type annotation for non-standard types
                sb.append("@Scalar<String>(\"${escapeString(value.toString())}\")")
            }
        }
    }

    private fun serializeArray(array: UDM.Array, sb: StringBuilder, depth: Int) {
        if (array.elements.isEmpty()) {
            sb.append("[]")
            return
        }

        sb.append("[")
        if (prettyPrint) sb.append("\n")

        array.elements.forEachIndexed { index, element ->
            if (prettyPrint) sb.append(indent(depth + 1))
            serializeValue(element, sb, depth + 1)
            if (index < array.elements.size - 1) {
                sb.append(",")
            }
            if (prettyPrint) sb.append("\n")
        }

        if (prettyPrint) sb.append(indent(depth))
        sb.append("]")
    }

    private fun serializeObject(obj: UDM.Object, sb: StringBuilder, depth: Int) {
        // Check if we need explicit @Object annotation
        val needsAnnotation = obj.name != null || obj.metadata.isNotEmpty()

        if (needsAnnotation) {
            sb.append("@Object")
            serializeObjectMeta(obj, sb, depth)
            sb.append(" ")
        }

        sb.append("{")
        if (prettyPrint) sb.append("\n")

        // Attributes section
        if (obj.attributes.isNotEmpty()) {
            if (prettyPrint) sb.append(indent(depth + 1))
            sb.append("attributes: {")
            if (prettyPrint) sb.append("\n")

            obj.attributes.entries.forEachIndexed { index, (key, value) ->
                if (prettyPrint) sb.append(indent(depth + 2))
                sb.append("$key: \"${escapeString(value)}\"")
                if (index < obj.attributes.size - 1) {
                    sb.append(",")
                }
                if (prettyPrint) sb.append("\n")
            }

            if (prettyPrint) sb.append(indent(depth + 1))
            sb.append("},")
            if (prettyPrint) sb.append("\n")
        }

        // Properties section
        if (obj.attributes.isNotEmpty() || needsAnnotation) {
            // Need explicit properties label
            if (prettyPrint) sb.append(indent(depth + 1))
            sb.append("properties: {")
            if (prettyPrint) sb.append("\n")
            serializeProperties(obj.properties, sb, depth + 2)
            if (prettyPrint) sb.append(indent(depth + 1))
            sb.append("}")
            if (prettyPrint) sb.append("\n")
        } else {
            // Shorthand: properties directly in object
            serializeProperties(obj.properties, sb, depth + 1)
        }

        if (prettyPrint) sb.append(indent(depth))
        sb.append("}")
    }

    private fun serializeObjectMeta(obj: UDM.Object, sb: StringBuilder, depth: Int) {
        if (obj.name == null && obj.metadata.isEmpty()) return

        sb.append("(")
        if (prettyPrint) sb.append("\n")

        val entries = mutableListOf<String>()

        // Name
        obj.name?.let {
            entries.add("${indent(depth + 1)}name: \"${escapeString(it)}\"")
        }

        // Metadata
        if (obj.metadata.isNotEmpty()) {
            val metadataStr = StringBuilder()
            metadataStr.append("${indent(depth + 1)}metadata: {")
            obj.metadata.entries.forEachIndexed { index, (key, value) ->
                metadataStr.append("$key: \"${escapeString(value)}\"")
                if (index < obj.metadata.size - 1) {
                    metadataStr.append(", ")
                }
            }
            metadataStr.append("}")
            entries.add(metadataStr.toString())
        }

        sb.append(entries.joinToString(",\n"))
        if (prettyPrint) sb.append("\n")
        if (prettyPrint) sb.append(indent(depth))
        sb.append(")")
    }

    private fun serializeProperties(properties: Map<String, UDM>, sb: StringBuilder, depth: Int) {
        properties.entries.forEachIndexed { index, (key, value) ->
            if (prettyPrint) sb.append(indent(depth))

            // Quote key if it contains special characters or is a reserved word
            val quotedKey = if (needsQuoting(key)) "\"${escapeString(key)}\"" else key
            sb.append("$quotedKey: ")

            serializeValue(value, sb, depth)

            if (index < properties.size - 1) {
                sb.append(",")
            }
            if (prettyPrint) sb.append("\n")
        }
    }

    private fun serializeBinary(binary: UDM.Binary, sb: StringBuilder) {
        // For now, just serialize size and indicate it's binary
        sb.append("@Binary(size: ${binary.data.size})")
        // TODO: Support base64 inline or external reference
    }

    private fun indent(depth: Int): String {
        return if (prettyPrint) " ".repeat(depth * indentSize) else ""
    }

    private fun escapeString(str: String): String {
        return str
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "\\r")
            .replace("\t", "\\t")
    }

    private fun needsQuoting(key: String): Boolean {
        // Quote if key contains special characters or starts with @
        return key.contains(Regex("[^a-zA-Z0-9_]")) || key.startsWith("@")
    }
}

/**
 * Extension function for convenient UDM serialization
 */
fun UDM.toUDMLanguage(prettyPrint: Boolean = true, sourceInfo: Map<String, String> = emptyMap()): String {
    return UDMLanguageSerializer(prettyPrint).serialize(this, sourceInfo)
}

File: theia-extension/utlx-theia-extension/src/browser/udm/udm-language-serializer.ts
/**
 * UDM Language Serializer
 *
 * Serializes UDM structures to UDM Language format (.udm files)
 * Ported from modules/core/src/main/kotlin/org/apache/utlx/core/udm/UDMLanguageSerializer.kt
 *
 * UDM Language is a meta-format that preserves complete UDM model state including:
 * - Type information (Scalar, Array, Object, DateTime, etc.)
 * - Metadata (source info, line numbers, validation state)
 * - Attributes (XML attributes, hints)
 * - Element names (XML context)
 *
 * This is different from standard YAML/JSON serialization which loses UDM metadata.
 *
 * Example:
 * ```typescript
 * const udm = UDMFactory.object(...);
 * const serializer = new UDMLanguageSerializer();
 * const udmLang = serializer.serialize(udm);
 * await fs.writeFile('output.udm', udmLang);
 * ```
 */

import { UDM, isScalar, isArray, isObject, isBinary, isDateTime, isDate, isLocalDateTime, isTime, isLambda } from './udm-core';

/**
 * Options for UDM Language serialization
 */
export interface SerializerOptions {
    prettyPrint?: boolean;
    indentSize?: number;
}

/**
 * Source information for UDM header
 */
export interface SourceInfo {
    source?: string;
    'parsed-at'?: string;
    [key: string]: string | undefined;
}

/**
 * UDM Language Serializer
 */
export class UDMLanguageSerializer {
    private readonly prettyPrint: boolean;
    private readonly indentSize: number;

    constructor(options: SerializerOptions = {}) {
        this.prettyPrint = options.prettyPrint ?? true;
        this.indentSize = options.indentSize ?? 2;
    }

    /**
     * Serialize UDM to UDM Language format
     */
    serialize(udm: UDM, sourceInfo: SourceInfo = {}): string {
        const sb: string[] = [];

        // Header
        sb.push('@udm-version: 1.0\n');
        if (Object.keys(sourceInfo).length > 0) {
            if (sourceInfo.source) {
                sb.push(`@source: "${sourceInfo.source}"\n`);
            }
            if (sourceInfo['parsed-at']) {
                sb.push(`@parsed-at: "${sourceInfo['parsed-at']}"\n`);
            }
        }
        sb.push('\n');

        // Body
        this.serializeValue(udm, sb, 0);

        return sb.join('');
    }

    private serializeValue(udm: UDM, sb: string[], depth: number): void {
        if (isScalar(udm)) {
            this.serializeScalar(udm, sb);
        } else if (isArray(udm)) {
            this.serializeArray(udm, sb, depth);
        } else if (isObject(udm)) {
            this.serializeObject(udm, sb, depth);
        } else if (isDateTime(udm)) {
            sb.push(`@DateTime("${udm.value}")`);
        } else if (isDate(udm)) {
            sb.push(`@Date("${udm.value}")`);
        } else if (isLocalDateTime(udm)) {
            sb.push(`@LocalDateTime("${udm.value}")`);
        } else if (isTime(udm)) {
            sb.push(`@Time("${udm.value}")`);
        } else if (isBinary(udm)) {
            this.serializeBinary(udm, sb);
        } else if (isLambda(udm)) {
            sb.push('@Lambda()');
        }
    }

    private serializeScalar(scalar: UDM & { type: 'scalar' }, sb: string[]): void {
        const value = scalar.value;

        if (value === null) {
            sb.push('null');
        } else if (typeof value === 'string') {
            sb.push(`"${this.escapeString(value)}"`);
        } else if (typeof value === 'boolean') {
            sb.push(value.toString());
        } else if (typeof value === 'number') {
            sb.push(value.toString());
        } else {
            // Explicit type annotation for non-standard types
            sb.push(`@Scalar<String>("${this.escapeString(String(value))}")`);
        }
    }

    private serializeArray(array: UDM & { type: 'array' }, sb: string[], depth: number): void {
        if (array.elements.length === 0) {
            sb.push('[]');
            return;
        }

        sb.push('[');
        if (this.prettyPrint) sb.push('\n');

        array.elements.forEach((element, index) => {
            if (this.prettyPrint) sb.push(this.indent(depth + 1));
            this.serializeValue(element, sb, depth + 1);
            if (index < array.elements.length - 1) {
                sb.push(',');
            }
            if (this.prettyPrint) sb.push('\n');
        });

        if (this.prettyPrint) sb.push(this.indent(depth));
        sb.push(']');
    }

    private serializeObject(obj: UDM & { type: 'object' }, sb: string[], depth: number): void {
        // Check if we need explicit @Object annotation
        const needsAnnotation = obj.name !== undefined || obj.metadata.size > 0;

        if (needsAnnotation) {
            sb.push('@Object');
            this.serializeObjectMeta(obj, sb, depth);
            sb.push(' ');
        }

        sb.push('{');
        if (this.prettyPrint) sb.push('\n');

        // Attributes section
        if (obj.attributes.size > 0) {
            if (this.prettyPrint) sb.push(this.indent(depth + 1));
            sb.push('attributes: {');
            if (this.prettyPrint) sb.push('\n');

            const attrEntries = Array.from(obj.attributes.entries());
            attrEntries.forEach(([key, value], index) => {
                if (this.prettyPrint) sb.push(this.indent(depth + 2));
                sb.push(`${key}: "${this.escapeString(value)}"`);
                if (index < attrEntries.length - 1) {
                    sb.push(',');
                }
                if (this.prettyPrint) sb.push('\n');
            });

            if (this.prettyPrint) sb.push(this.indent(depth + 1));
            sb.push('},');
            if (this.prettyPrint) sb.push('\n');
        }

        // Properties section
        if (obj.attributes.size > 0 || needsAnnotation) {
            // Need explicit properties label
            if (this.prettyPrint) sb.push(this.indent(depth + 1));
            sb.push('properties: {');
            if (this.prettyPrint) sb.push('\n');
            this.serializeProperties(obj.properties, sb, depth + 2);
            if (this.prettyPrint) sb.push(this.indent(depth + 1));
            sb.push('}');
            if (this.prettyPrint) sb.push('\n');
        } else {
            // Shorthand: properties directly in object
            this.serializeProperties(obj.properties, sb, depth + 1);
        }

        if (this.prettyPrint) sb.push(this.indent(depth));
        sb.push('}');
    }

    private serializeObjectMeta(obj: UDM & { type: 'object' }, sb: string[], depth: number): void {
        if (obj.name === undefined && obj.metadata.size === 0) return;

        sb.push('(');
        if (this.prettyPrint) sb.push('\n');

        const entries: string[] = [];

        // Name
        if (obj.name !== undefined) {
            entries.push(`${this.indent(depth + 1)}name: "${this.escapeString(obj.name)}"`);
        }

        // Metadata
        if (obj.metadata.size > 0) {
            const metadataParts: string[] = [];
            metadataParts.push(`${this.indent(depth + 1)}metadata:`);

            if (this.prettyPrint) {
                metadataParts.push('\n');
                metadataParts.push(`${this.indent(depth + 1)}{`);
                metadataParts.push('\n');

                const metaEntries = Array.from(obj.metadata.entries());
                metaEntries.forEach(([key, value], index) => {
                    metadataParts.push(`${this.indent(depth + 2)}${key}: "${this.escapeString(value)}"`);
                    if (index < metaEntries.length - 1) {
                        metadataParts.push(',');
                    }
                    metadataParts.push('\n');
                });

                metadataParts.push(`${this.indent(depth + 1)}}`);
            } else {
                metadataParts.push(' {');
                const metaEntries = Array.from(obj.metadata.entries());
                metaEntries.forEach(([key, value], index) => {
                    metadataParts.push(`${key}: "${this.escapeString(value)}"`);
                    if (index < metaEntries.length - 1) {
                        metadataParts.push(', ');
                    }
                });
                metadataParts.push('}');
            }

            entries.push(metadataParts.join(''));
        }

        sb.push(entries.join(',\n'));
        if (this.prettyPrint) sb.push('\n');
        if (this.prettyPrint) sb.push(this.indent(depth));
        sb.push(')');
    }

    private serializeProperties(properties: Map<string, UDM>, sb: string[], depth: number): void {
        const propEntries = Array.from(properties.entries());
        propEntries.forEach(([key, value], index) => {
            if (this.prettyPrint) sb.push(this.indent(depth));

            // Quote key if it contains special characters or is a reserved word
            const quotedKey = this.needsQuoting(key) ? `"${this.escapeString(key)}"` : key;
            sb.push(`${quotedKey}: `);

            this.serializeValue(value, sb, depth);

            if (index < propEntries.length - 1) {
                sb.push(',');
            }
            if (this.prettyPrint) sb.push('\n');
        });
    }

    private serializeBinary(binary: UDM & { type: 'binary' }, sb: string[]): void {
        // For now, just serialize size and indicate it's binary
        sb.push(`@Binary(size: ${binary.data.length})`);
        // TODO: Support base64 inline or external reference
    }

    private indent(depth: number): string {
        return this.prettyPrint ? ' '.repeat(depth * this.indentSize) : '';
    }

    private escapeString(str: string): string {
        return str
            .replace(/\\/g, '\\\\')
            .replace(/"/g, '\\"')
            .replace(/\n/g, '\\n')
            .replace(/\r/g, '\\r')
            .replace(/\t/g, '\\t');
    }

    private needsQuoting(key: string): boolean {
        // Quote if key contains special characters or starts with @
        return /[^a-zA-Z0-9_]/.test(key) || key.startsWith('@');
    }
}

/**
 * Convenience function for serializing UDM
 */
export function toUDMLanguage(udm: UDM, prettyPrint: boolean = true, sourceInfo: SourceInfo = {}): string {
    const serializer = new UDMLanguageSerializer({ prettyPrint });
    return serializer.serialize(udm, sourceInfo);
}
